autopilot/manager.go

```diff
@@ -0,0 +1,277 @@
+
+	// SubscribeTopology is used to get a subscription for topology changes
+	// on the network.
+	SubscribeTopology func() (*routing.TopologyClient, error)
+}
+
+// Manager is struct that manages an autopilot agent, making it possible to
+// enable and disable it at will, and hand it relevant external information.
+// It implements the autopilot grpc service, which is used to get data about
+// the running autopilot, and give it relevant information.
+type Manager struct {
+	started uint32 // To be used atomically.
+	stopped uint32 // To be used atomically.
+
+	cfg *ManagerCfg
+
+	// pilot is the current autopilot agent. It will be nil if the agent is
+	// disabled.
+	pilot *Agent
+
+	quit chan struct{}
+	wg   sync.WaitGroup
+	sync.Mutex
+}
+
+// NewManager creates a new instance of the Manager from the passed config.
+func NewManager(cfg *ManagerCfg) (*Manager, error) {
+	return &Manager{
+		cfg:  cfg,
+		quit: make(chan struct{}),
+	}, nil
+}
+
+// Start starts the Manager.
+func (r *Manager) Start() error {
+	if !atomic.CompareAndSwapUint32(&r.started, 0, 1) {
+		return nil
+	}
+
+	return nil
+}
+
+// Stop stops the Manager. If an autopilot agent is active, it will also be
+// stopped.
+func (r *Manager) Stop() error {
+	if !atomic.CompareAndSwapUint32(&r.stopped, 0, 1) {
+		return nil
+	}
+
+	if err := r.StopAgent(); err != nil {
+		log.Errorf("Unable to stop pilot: %v", err)
+	}
+
+	close(r.quit)
+	r.wg.Wait()
+
+	return nil
+}
+
+// GetStatus returns the current active status of the autopilot agent.
+func (r *Manager) GetStatus() bool {
+	r.Lock()
+	defer r.Unlock()
+
+	return r.pilot != nil
+}
+
+// StartAgent creates and starts an autopilot agent from the Manager's
+// config.
+func (r *Manager) StartAgent() error {
+	r.Lock()
+	defer r.Unlock()
+
+	// Already active.
+	if r.pilot != nil {
+		return nil
+	}
+
+	// Next, we'll fetch the current state of open channels from the
+	// database to use as initial state for the auto-pilot agent.
+	initialChanState, err := r.cfg.ChannelState()
+	if err != nil {
+		return err
+	}
+
+	// Now that we have all the initial dependencies, we can create the
+	// auto-pilot instance itself.
+	r.pilot, err = New(*r.cfg.PilotCfg, initialChanState)
+	if err != nil {
+		return err
+	}
+
+	if err := r.pilot.Start(); err != nil {
+		return err
+	}
+
+	// Finally, we'll need to subscribe to two things: incoming
+	// transactions that modify the wallet's balance, and also any graph
+	// topology updates.
+	txnSubscription, err := r.cfg.SubscribeTransactions()
+	if err != nil {
+		defer r.pilot.Stop()
+		return err
+	}
+	graphSubscription, err := r.cfg.SubscribeTopology()
+	if err != nil {
+		defer r.pilot.Stop()
+		defer txnSubscription.Cancel()
+		return err
+	}
+
+	// We'll launch a goroutine to provide the agent with notifications
+	// whenever the balance of the wallet changes.
+	// TODO(halseth): can lead to panic if in process of shutting down.
```

Comment:
What leads to a panic? The `wg`?

Reply:
Yes, as @cfromknecht has mentioned earlier, this could lead to panic in certain very rare cases.